### PR TITLE
Clean reconstructed objects outside pack window

### PR DESCRIPTION
**Pull request by jfontan**
Object walk reconstructs delta objects but these are not cleaned up
after they got out the pack window. Without this change all
reconstructed objects reside in memory.

restoreOriginal call is moved before calling Size(). Now we can not
guarantee that the object is already undeltified.

### CODE REVIEW

File: plumbing/format/packfile/delta_selector.go [hunk 1]
```diff
@@ -261,6 +267,16 @@ func (dw *deltaSelector) walk(
 }
 
 func (dw *deltaSelector) tryToDeltify(indexMap map[plumbing.Hash]*deltaIndex, base, target *ObjectToPack) error {
+	// Original object might not be present if we're reusing a delta, so we
+	// ensure it is restored.
+	if err := dw.restoreOriginal(target); err != nil {
```
Comment (erizocosmico) [inline plumbing/format/packfile/delta_selector.go]: isn't this making it actually slower than before (even if it uses way less memory)?

Comment (jfontan) [reply plumbing/format/packfile/delta_selector.go]: I've just run `borges pack` with repository https://github.com/numpy/numpy (downloaded locally) and the results are quite close.

* Original version:
  * 203.83user 12.34system 2:28.80elapsed 2.4 Gb RAM
  * 204.40user 12.27system 2:29.30elapsed 2.4 Gb RAM
* With proposed changes:
  * 217.82user 11.97system 2:25.72elapsed 1 Gb RAM
  * 220.06user 11.81system 2:26.28elapsed 1 Gb RAM

It seems that it's a bit slower (user time is bigger) but it spends more system time. Maybe heap allocation.

**Comment by erizocosmico** [comment]
Then lgtm. That little slowdown is a nice tradeoff for half the memory
On Thu, 11 Jan 2018 at 19:19, Javi Fontan <notifications@github.com<mailto:notifications@github.com>> wrote:

@jfontan commented on this pull request.

________________________________

In plumbing/format/packfile/delta_selector.go<https://github.com/src-d/go-git/pull/716#discussion_r161034780>:

> @@ -261,6 +267,16 @@ func (dw *deltaSelector) walk(
 }

 func (dw *deltaSelector) tryToDeltify(indexMap map[plumbing.Hash]*deltaIndex, base, target *ObjectToPack) error {
+       // Original object might not be present if we're reusing a delta, so we
+       // ensure it is restored.
+       if err := dw.restoreOriginal(target); err != nil {


I've just run borges pack with repository https://github.com/numpy/numpy (downloaded locally) and the results are quite close.

  *   Original version:
     *   203.83user 12.34system 2:28.80elapsed 2.4 Gb RAM
     *   204.40user 12.27system 2:29.30elapsed 2.4 Gb RAM
  *   With proposed changes:
     *   217.82user 11.97system 2:25.72elapsed 1 Gb RAM
     *   220.06user 11.81system 2:26.28elapsed 1 Gb RAM

It seems that it's a bit slower (user time is bigger) but it spends more system time. Maybe heap allocation.

**Comment by mcuadros** [comment]
Can we test this with more diverse repository? Not only with the `numpy` repository?

**Comment by jfontan** [comment]
I've did new tests with the following repos:

* **upsilon**: small repository, https://github.com/numpy/numpy
* **numpy**: average repository, https://github.com/numpy/numpy
* **tensorflow**, average repository, https://github.com/tensorflow/tensorflow
* **bismuth**, some files are 100Mb in size, https://github.com/hclivess/Bismuth

The times and memory are only from the push action from local to a new repository. Each test is executed twice and the smaller value for time and memory is selected.

| repository | master | fix memory |
| --------------|-----------|----------------|
| upsilon | 54.2s 387MiB | 53.4s 387MiB |
| numpy | 1m25s 1.9GiB | 1m19s 0.65GiB |
| tensorflow | 3m16s 3.9GiB | 2m59s 1.28GiB |
| bismuth | 12m16s 17.7GiB | 10m37s 3.13GiB |

The code to do the benchmark is here: https://gist.github.com/jfontan/42fbfe0761e5280012d285c1a4290f5d

**Comment by erizocosmico** [comment]
So it's actually faster? And uses half the memory? Nice, great job!

**Comment by mcuadros** [comment]
Networking is envolved, so speed number aren't reliable

**Comment by jfontan** [comment]
I've just updated the table with two small repos.

**Pull request merged**